### PR TITLE
feat: highlight OV menu - CXSPA-1440

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/components/overview-form/configurator-overview-form.component.html [hunk 1]
```diff
@@ -80,7 +83,7 @@ <h2 aria-hidden="true" id="{{ getGroupId(idPrefix, group.id) }}">
             context: {
               overviewGroups: group.subGroups,
               level: level + 1,
-              idPrefix: idPrefix + group.id
+              idPrefix: idPrefix ? idPrefix + '--' + group.id : group.id
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/overview-form/configurator-overview-form.component.html]: Why was that needed (just because I am curious)?

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html [hunk 2]
```diff
@@ -17,31 +17,34 @@
...
-      <ng-container *ngIf="group.subGroups?.length > 0">
-        <ng-container
-          *ngTemplateOutlet="
-            groups;
-            context: {
-              overviewGroups: group.subGroups,
-              level: level + 1,
-              idPrefix: idPrefix + group.id
-            }
+  <ul>
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html:20]: This breaking change needs to be documented for customers

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts [hunk 3]
```diff
@@ -113,15 +176,271 @@ describe('ConfigurationOverviewMenuComponent', () => {
...
+
+  describe('getGroupId', () => {
+    it('should dispatch request to utils service', () => {
+      initialize();
+      expect(component.getGroupId('A', 'B')).toBe('A--B-ovGroup');
+    });
+
+    it('should cope with undefined idPrefix', () => {
+      initialize();
+      expect(component.getGroupId(undefined, 'B')).toBe('B-ovGroup');
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts]: Strict mode issue. 
Please fix all strict mode issues in this test. Please consider to move to VSCode (there you see them directly), we are facing this issue very often

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts [hunk 4]
```diff
@@ -13,35 +21,156 @@ import { ConfiguratorStorefrontUtilsService } from '../service/configurator-stor
...
    */
   navigateToGroup(prefix: string, id: string): void {
     const ovGroupId = this.configuratorStorefrontUtilsService.createOvGroupId(
       prefix,
       id
     );
+
     this.configuratorStorefrontUtilsService.scrollToConfigurationElement(
-      '#' + ovGroupId
+      '#' + ovGroupId + ' h2'
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts:144]: Why that change? Do we need to document it?

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 5]
```diff
@@ -223,8 +227,26 @@ describe('ConfigUtilsService', () => {
   });
 
   describe('createOvGroupId', () => {
+    it('should create a group id from its 1 parameter', () => {
+      expect(classUnderTest.createOvGroupId(undefined, 'B')).toBe('B-ovGroup');
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]: Strict mode issue. Please remove all strict mode issues in this test class!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 6]
```diff
@@ -254,7 +254,21 @@ export class ConfiguratorStorefrontUtilsService {
    * @returns {string} - generated group ID
    */
   createOvGroupId(prefix: string, groupId: string): string {
-    return `id${prefix}${groupId}-ovGroup`;
+    return prefix ? prefix + '--' + groupId + '-ovGroup' : groupId + '-ovGroup';
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: We need to document the changed behaviour

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 7]
```diff
@@ -297,4 +311,122 @@ export class ConfiguratorStorefrontUtilsService {
...
+    return 0;
+  }
+
+  /**
+   * Syncs scrolling.
+   *
+   * @param {string} querySelector - Element query selector
+   * @param {HTMLElement | undefined} element - Element that should be visible within the scrollable element.
+   */
+  syncScroll(querySelector: string, element: HTMLElement | undefined): void {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Please find a better name for this one, what exactly is its purpose?

File: feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss [hunk 8]
```diff
@@ -1,34 +1,61 @@
...
-        border: none;
-        background-color: transparent;
-
-        &:hover {
-          cursor: pointer;
-          color: var(--cx-color-primary);
-        }
+    overflow-y: auto;
+
+    ul {
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss:5]: Document breaking changes

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 7 above)
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Please also try to improve the test coverage of this class. Statements and branch coverage is down significantly

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 9]
```diff
@@ -223,8 +227,26 @@ describe('ConfigUtilsService', () => {
   });
 
   describe('createOvGroupId', () => {
+    it('should create a group id from its 1 parameter', () => {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]:
```suggestion
    it('should create a group id for a top level group without prefix', () => {
```

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 10]
```diff
@@ -223,8 +227,26 @@ describe('ConfigUtilsService', () => {
...
+      expect(classUnderTest.createOvGroupId(undefined, 'B')).toBe('B-ovGroup');
+    });
+
     it('should create a group id from its 2 parameters', () => {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]:
```suggestion
    it('should create a group id for a child group with prefix, () => {
```

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 11]
```diff
@@ -223,8 +227,26 @@ describe('ConfigUtilsService', () => {
...
-      expect(classUnderTest.createOvGroupId('A', 'B')).toBe('idAB-ovGroup');
+      expect(classUnderTest.createOvGroupId('A', 'B')).toBe('A--B-ovGroup');
+    });
+  });
+
+  describe('createOvMenuItemId', () => {
+    it('should create a menu item id from its 1 parameter', () => {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]:
```suggestion
    it('should create a menu item id for a top level group without prefix', () => {
```

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 12]
```diff
@@ -223,8 +227,26 @@ describe('ConfigUtilsService', () => {
...
+      expect(classUnderTest.createOvMenuItemId(undefined, 'B')).toBe(
+        'B-ovMenuItem'
+      );
+    });
+
+    it('should create a menu item id from its 2 parameters', () => {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]:
```suggestion
    it('should create a menu item id for a child group with prefix', () => {
```

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts [hunk 13]
```diff
@@ -303,6 +325,25 @@ describe('ConfigUtilsService', () => {
...
+
+        focusedElements.forEach((focusedElement, index) => {
+          if (index === focusedElementIndex) {
+            expect(focusedElement.focus).toHaveBeenCalled();
+          } else {
+            expect(focusedElement.focus).not.toHaveBeenCalled();
+          }
+        });
+      }
+
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]: thx - good refactoring :-)

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts [hunk 14]
```diff
@@ -46,47 +72,84 @@ function initialize() {
...
+  afterEach(() => {
+    document.body.removeChild(htmlElem);
+  });
+
   it('should create component', () => {
     initialize();
     expect(component).toBeDefined();
   });
 
+  it('should call ngAfterViewInit', () => {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts]: ngAfterViewInit is the method under test? but what is expected here? The name should be something like
'on ngAfterViewInit something should happen'

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts:144]: I don't think that it should be documented. 
The reason of this change is that it will be scrolled to the corresponding group header and not to the group container.

File: feature-libs/product-configurator/rulebased/components/overview-form/configurator-overview-form.component.html (hunk 1 above)
Comment (Uli-Tiger) [reply feature-libs/product-configurator/rulebased/components/overview-form/configurator-overview-form.component.html]: can we move this logic to configurator-storefront-utils, as well?
it duplicated inn the configurator-overview-menu component. Also the seperator
'--' is hard coded in several locations. Would be easier if all is encapsulated in one service class (configurator-storefront-utils)

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-form/configurator-overview-form.component.html]: In order to better map a group to the corresponding menu item, for instance:
**group:** KD990WRES--KD990WRESP--KD990WRESPS--KD990WRESP1--TECH**-ovGroup**
**menu item:** KD990WRES--KD990WRESP--KD990WRESPS--KD990WRESP1--TECH**-ovMenuItem**

We have to only cut a suffix **-ovGroup** und replace by **-ovMenuItem**

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts (hunk 3 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html [hunk 15]
```diff
@@ -17,31 +17,34 @@
...
+          <cx-icon [type]="iconTypes.ARROW_LEFT" aria-hidden="true"></cx-icon>
+        </button>
+        <ng-container *ngIf="group.subGroups?.length > 0">
+          <ng-container
+            *ngTemplateOutlet="
+              groups;
+              context: {
+                overviewGroups: group.subGroups,
+                level: level + 1,
+                idPrefix: idPrefix ? idPrefix + '--' + group.id : group.id
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html]: line is duplicated in configurator-oveview maybe we can move logic to configurator-storefront-utils

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 16]
```diff
@@ -297,4 +311,122 @@ export class ConfiguratorStorefrontUtilsService {
...
+   * @returns {number} - Height of the spare viewport.
+   */
+  getSpareViewportHeight(): number {
+    if (this.windowRef.isBrowser()) {
+      const spaHeaderHeight = this.getHeight('header');
+      const ovHeaderHeight = this.getHeight('.VariantConfigOverviewHeader');
+      const addToCartHeight =
+        this.getHeight('cx-configurator-add-to-cart-button') !== 0
+          ? this.getHeight('cx-configurator-add-to-cart-button')
+          : 82;
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: magic number  - Please replace 82 by a const with a descriptive name

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts [hunk 17]
```diff
@@ -297,4 +311,122 @@ export class ConfiguratorStorefrontUtilsService {
...
   }
+
+  /**
+   * Retrieves a list of HTML elements based on querySelector when running in browser
+   *
+   * @param {string} querySelector - querySelector
+   * @returns {HTMLElement[] | undefined} - List of HTML elements
+   */
+  getElements(querySelector: string): HTMLElement[] | undefined {
+    if (this.windowRef.isBrowser()) {
```
Comment (Uli-Tiger) [inline feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts:334]: do we need this safeguard at all?
What is the use case, where this gets executed and we are NOT in the browser?

**Review by Uli-Tiger — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: I even talked to Sebastian to find a better name but he did not have a better solution as it is.
Do you have a better name for this method?

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts (hunk 10 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html (hunk 2 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html:20]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts (hunk 5 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.ts:144]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss (hunk 8 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss:5]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts (hunk 13 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.spec.ts]: You are welcome!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 16 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html (hunk 15 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html]: It's true but it is how the context of the template is normally defined.
The question is whether it will be more readable if the context will be extracted to the helper class.
I don't think so then one has to jump to the helper class to better understand what context will be forwarded at this play.
I would suggest to let it be as it is.

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts (hunk 14 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.spec.ts]: The test name should start with `should`
I changed the name to `should call ngAfterViewInit after ovMenu is rendered`

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts]: Done!

File: feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss (hunk 8 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/styles/_configurator-overview-menu.scss:5]: Done!

File: feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts (hunk 17 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts:334]: As far as I understand by means of this condition `this.windowRef.isBrowser()`, we want to guarantee that this method is called only for the browser

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts:334]: If we remove this condition here, then the customer will ask him/herself why in other methods we still have this condition.

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/service/configurator-storefront-utils.service.ts:334]: I strongly believe that this condition does not have any impact on the performance.

File: feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html (hunk 15 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html]: I extracted to a separate method `getPrefixId(idPrefix, group.id)`

Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/overview-menu/configurator-overview-menu.component.html]: What about it?

**Review by Uli-Tiger — DISMISSED** [review]

File: docs/migration/6_0/typescript-manual-doc.md [hunk 18]
```diff
@@ -79,6 +79,14 @@ If explicitly set, this option will take precedence over the express server.
   is much smaller and includes `configurator-group.component.html`, moreover many methods previously residing in `configurator-form.component.ts` have
   been moved to `configurator-group.component.ts`
 
+### ConfiguratorOverviewMenuComponent
+
+- By navigation to a certain overview group via overview menu the browser does not scroll anymore to the container of the group  `'#' + ovGroupId`, namely to the title of the corresponding group `'#' + ovGroupId+ ' h2'`.
```
Comment (ChristophHi) [inline docs/migration/6_0/typescript-manual-doc.md]:
```suggestion
- By navigation to a certain overview group via overview menu the browser does not scroll anymore to the container of the group  `'#' + ovGroupId`, but to the title of the corresponding group `'#' + ovGroupId+ ' h2'`.
```

**Review by ChristophHi — CHANGES_REQUESTED** [review]

**Review by ChristophHi — APPROVED** [review]

**Review by Uli-Tiger — APPROVED** [review]

**Review by hackergil — APPROVED** [review]
